### ISSUE TITLE
Closing of the pipe on String or Bytes?

### ISSUE THREAD

**Issue by lobre** [bitfield/script#41]
Hello,

First, thanks for this project. Really like the API and design!

Instead of an issue, I will more ask a question. I am not sure to understand the strategy of closing for a Pipe when the sink is `String()` or `Bytes()`.

For `String()` (https://github.com/bitfield/script/blob/master/sinks.go#L65), the documentation says: 
> String returns the contents of the Pipe as a string, or an error, and closes the pipe after reading

However, I cannot see in the code any reference to a close method. It only calls `Bytes()` which itself does not seem to close the reader. Am I missing something? Or is it a documentation mistake?

I can see in a previous commit that the close has been deleted. Is it on purpose?

https://github.com/bitfield/script/commit/062e038e4a2e47342a0798a347fe2bdf57884c88#diff-1d716f4c8f87036767ecdd0412ad974fL30

Thanks for the clarification!
Loric

**Comment by bitfield** [comment]
Hey @lobre! Thanks for stopping by. Nice to meet you, and thanks for the kind words.

> I cannot see in the code any reference to a close method. It only calls Bytes() which itself does not seem to close the reader.

That's a great question. I had to think about this a little when coming up with the original design for pipes. My first thought was simply to close them in any method which reads a pipe completely, such as `String()`. But you have to remember to do that every time, and never make a mistake, or you'll end up leaking files. Also, a pipe itself is an `io.Reader`, so users can create pipes and read from them, and _they_ would have to remember to close the pipes, too. I don't like patterns where if you forget to do something, there are bad consequences. It's too fragile.

So pipes actually wrap `io.Reader` in a type called `ReadAutoCloser`, which ensures that whatever the source of the pipe is, it's always closed when the pipe is read to completion. That's very simple to do:

```go
// Read reads up to len(p) bytes from the data source into p. It returns the
// number of bytes read and any error encountered. At end of file, Read returns
// 0, io.EOF. In the EOF case, the data source will be closed.
func (a ReadAutoCloser) Read(b []byte) (n int, err error) {
	if a.r == nil {
		return 0, io.EOF
	}
	n, err = a.r.Read(b)
	if err == io.EOF {
		a.Close()
	}
	return n, err
}
```

This is much nicer, as you _can_ close the pipe if you want to, but if you don't, it will always be closed once you've read everything from it. (You still have to read everything from it, and this is called out in the [Closing Pipes](https://github.com/bitfield/script#closing-pipes) section of the documentation, but this is better than having to read _and_ close it.)

Incidentally, it's always safe to close a pipe, even if the source it's reading from is not an `io.Closer`. To ensure this, `NewReadAutoCloser()` detects if the source is not a `Closer` and wraps it in an `ioutil.NopCloser`.

To make sure this doesn't accidentally get broken somehow, many of the tests do things which would read the pipe completely, and then check that the source was closed:

```go
	input := File("testdata/first.input.txt")
	got, err := input.First(10).Bytes()
	...
	_, err = ioutil.ReadAll(input.Reader)
	if err == nil {
		t.Error("input not closed after reading")
	}
```

Maybe this isn't the smartest way to do it, but it's the way I thought of! I'd love to hear any suggestions for improving it, or improving the library generally. Do feel free to open an issue for anything you find.

I hope this answers your excellent question, but if not, let me know.

**Comment by lobre** [comment]
Hi @bitfield,

Thanks for the quick and precise answer. I did not know this pattern and I did not realise this `a.Close()` in the `Read` method! It makes definitely more sense now.

And I think this pattern is rather interesting even if I haven't seen it anywhere. As you said, it is a good way to avoid thinking about closing pipes! And I think it makes even more sense in an application such as `script` that operates in a "functional" manner. The aim of this project is to quickly script the operating system. Having to always close pipes could be a burden going in the opposite direction.

Tks,
Loric
